### ISSUE TITLE
Misleading diagnostic when Deref is implemented. [E0308]

### ISSUE THREAD

**Issue by JCBurnside** [rust-lang/rust#83563]
Given the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=99918b2dd711e6d34f9d0345873bb616
```rust
use std::ops::Deref;

struct Foo<'a>(&'a str);

impl <'a> Deref for Foo<'a> {
    type Target = &'a str;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

fn fun(s:&str){
    println!("{}",s);
}

fn main() {

    let foo = Foo("x");
    fun(foo);
}
```

The current output is:

```
error[E0308]: mismatched types
  --> src/lib.rs:19:9
   |
19 |     fun(foo);
   |         ^^^
   |         |
   |         expected `&str`, found struct `Foo`
   |         help: consider borrowing here: `&foo`
```

How ever I feel the term borrowing in the diagnostic is misleading since we are not borrowing `foo` but rather dereferencing it.

**Comment by osa1** [comment]
> we are not borrowing foo but rather dereferencing it.

You borrow it, that's what `&` does. Dereferencing is done by the compiler. So in that sense the diagnostic is correct.

How would you tell the user to pass borrow of `foo` here?

**Comment by estebank** [comment]
Current output:
```
error[E0308]: mismatched types
  --> src/main.rs:19:9
   |
19 |     fun(foo);
   |     --- ^^^ expected `&str`, found `Foo<'_>`
   |     |
   |     arguments to this function are incorrect
   |
note: function defined here
  --> src/main.rs:12:4
   |
12 | fn fun(s:&str){
   |    ^^^ ------
help: consider borrowing here
   |
19 |     fun(&foo);
   |         +
```

The only change that would be useful here would be to change the wording to be
```
help: consider borrowing here to dereference `Foo<'_>` to `&str`
```
